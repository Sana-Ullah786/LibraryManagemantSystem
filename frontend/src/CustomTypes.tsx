/**
 * This file contains custom data types and interfaces that are generic enough to be used multiple times throughout the project
 */

export interface AuthorDetails {
  first_name?: string | null;
  last_name?: string | null;
  date_of_birth?: string | null;
  date_of_death?: string | null;
}

export interface Author extends AuthorDetails {
  id: number;
}

export type UserLoginData = {
  username: string;
  password: string;
};

export type Tokens = {
  access_token: string;
  refresh_token: string;
};

// The attributes of a book
export interface BookBase {
  title: string;
  description: string;
  isbn: string;
  dateOfPublication: Date;
}

export interface BookOut extends BookBase {
  authorIds: number[];
  genreIds: number[];
  languageId: number;
  numberOfCopies: number;
}

export interface BookSaved extends BookBase {
  id: number;
  authors: Author[];
  language: Language;
  genres: Genre[];
}

//Genre Interfaces


export interface Genre {
  id: number;
  genre: string ;
}

export interface GenreDetails {
  id: number;
  genre: string;
}

export interface ErrorObject {
  status: number;
  message: string;
}

export interface GenreMap {
  [id: number]: Genre;
}

export interface Language {
  id: number;
  language: string;
}
//Signup Attrubutes
export interface SignupData {
  email: string;
  username: string;
  password: string;
  first_name: string;
  last_name: string;
  contact_number: string;
  address: string;
}

//Language Attributes

export interface LanguageDetails {
  language: string | null;
  // Add other properties specific to the LanguageDetails type
}
export interface Language {
  id: number;
  language: string;
}
