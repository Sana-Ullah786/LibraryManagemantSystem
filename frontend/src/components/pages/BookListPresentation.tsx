import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import { BookSaved } from '../../CustomTypes';
import LibrarianLinks from '../LibrarianLinks';
import '../style.css'; // Import the Genres CSS file
import ScrollView from '../Scrollview';

export const BookListPresentation = (props: {
  books: BookSaved[];
  url: string;
  showLinks: boolean;
}) => {
  const [searchTerm, setSearchTerm] = useState<string>(''); // State for the search term

  // Filtered books based on search term
  const filteredBooks = props.books.filter(book =>
    book.title.toLowerCase().includes(searchTerm.toLowerCase())
  );

  const handleSearchChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setSearchTerm(e.target.value);
  };

  return (
    <div className='modal'>
      <div style={{ display: 'flex', justifyContent: 'center' }}>
        <h1>Books</h1>
      </div>
      <div style={{ display: 'flex', flexDirection: 'row', justifyContent: 'space-evenly' }}>
        <div>
          <h3>Search Book: </h3>
        </div>
        <div style={{ marginTop: 18 }}>
          <input
            type="text"
            placeholder="Search books..."
            value={searchTerm}
            onChange={handleSearchChange}
          />
        </div>
      </div>
      <br />
      <ScrollView>
        <div className="card-container">
          {filteredBooks.map((book) => (
            <div key={book.id.toString()} className="card" data-testid={"item" + book.id.toString()}>
              <Link to={"/books/" + book.id.toString()}>{book.title}</Link>
              <p><b>Authors:</b></p>
              <p>
              {
                
                book.authors.map((author) => {
                  return (
                   ` ${author.first_name} ${author.last_name}`
                  )
                }).join(", ")}
</p>
<p><b>Description :</b></p>
              <p>{book.description}</p>

              {props.showLinks && (
                <LibrarianLinks url={`${props.url}/${book.id}`} />
              )}
            </div>
          ))}
        </div>
      </ScrollView>
      {props.showLinks && (
        <div style={{ display: 'flex', justifyContent: 'center' }}>
          <div>
            <Link to={`${props.url}/create`} className='genre-link'> Create Book </Link>
          </div>
        </div>
      )}
    </div>
  );
};
