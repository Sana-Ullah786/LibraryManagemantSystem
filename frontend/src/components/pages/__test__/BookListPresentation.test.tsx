import { render, cleanup } from '@testing-library/react';
import { BookListPresentation } from '../BookListPresentation';
import { Book, AuthorAttributes, Genre, Language } from '../../../CustomTypes';
import { StaticRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';

//TODO: changes to the custom types file means these test cases won't work anymore and will need to be changed
const books: Book[] = [
    {
        id: 1,
        title: 'title1',
        authorId: 1,
        summary: 'summary',
        isbn: '1231231231231',
        languageId: 1,
        genreIds: [1, 2]
    },
    {
        id: 2,
        title: 'title2',
        authorId: 1,
        summary: 'summary',
        isbn: '1231231231231',
        languageId: 1,
        genreIds: [2]
    }
];

const author: AuthorAttributes = {
    id: 1,
    first_name: 'firstname',
    last_name: 'lastname',
    date_of_birth: 'dateofbirth',
    date_of_death: 'dateofdeath' 
};

const language: Language = {
    id: 1,
    name: 'English'
};

const genres: Genre[] = [
    {
        id: 1,
        name: 'Fantasy'
    },
    {
        id: 2,
        name: 'Drama'
    }
];

afterEach(cleanup);


describe("renders correct content", () => {

    /**
     * This test checks if the component renders without any errors.
     */
    test('renders without error', () => {

        render(
            < StaticRouter >
                <BookListPresentation books={ books } url={ "test-url" }/>
            </StaticRouter>
        );

    });

    /**
     * This test checks that each list item contains a Link element
     */
    test("attaches a link to each list item", () => {

        const { getByTestId } = render(
            < StaticRouter >
                <BookListPresentation books={ books } url={ "test-url" }/>
            </StaticRouter>
        );

        for(const book of books){
            expect(getByTestId("item" + book.id).firstElementChild?.tagName).toBe("A");
        }

    });

    /**
     * This test checks that each list item contains the correct url
     */
     test("attaches correct url to each link", () => {

        const { getByTestId } = render(
            < StaticRouter >
                <BookListPresentation books={ books } url={ "test-url" }/>
            </StaticRouter>
        );

        for(const book of books){
            expect(getByTestId("item" + book.id).firstElementChild).toHaveAttribute("href", "/test-url/" + book.id);
        }

    });

});


describe("tests correct data displayed", () => {

    /**
    * This test checks that correct text is displayed for each item
    */
    test("correct text displayed", () => {

        const { getByTestId } = render(
            < StaticRouter >
                <BookListPresentation books={ books } url={ "test-url" }/>
            </StaticRouter>
        );

        for(const book of books){
            expect(getByTestId("item" + book.id)).toHaveTextContent(book.title);
        }

    });

});

describe("Snapshot tests", () => {

    /**
     * This test makes sure that the snapshot produced is valid and up to date.
     * It detects any changes in user interface and reports them.
     */
    test("produces correct snapshot", () => {

        const tree = renderer
            .create(
                < StaticRouter >
                    <BookListPresentation books={ books } url={ "test-url" }/>
                </StaticRouter>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();

    });

});