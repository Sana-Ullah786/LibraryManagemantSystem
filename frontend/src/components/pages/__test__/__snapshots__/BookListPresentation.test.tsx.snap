// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot tests produces correct snapshot 1`] = `
<div>
  <ul>
    <li
      data-testid="item1"
    >
      <a
        href="/test-url/1"
        onClick={[Function]}
      >
        title1
      </a>
    </li>
    <li
      data-testid="item2"
    >
      <a
        href="/test-url/2"
        onClick={[Function]}
      >
        title2
      </a>
    </li>
  </ul>
</div>
`;
